File: TafelbildiOS/Tafelbild/Model/Zuruecksetzen.swift
import Foundation

/// Elemente wieder auf „unbenutzt" stellen.
///
/// **Warum das fehlte.** Alles, was auf der Tafel abläuft, bleibt danach
/// stehen: der zuletzt gezogene Name, die ausgeloste Sitzordnung, die
/// gelaufene Geburtstagsfeier, die abgehakten Punkte des Tagesablaufs. Das
/// ist im Unterricht richtig — eine Tafel, die sich selbst aufräumt,
/// während die Klasse noch hinsieht, wäre unbrauchbar. Am nächsten Morgen
/// ist es falsch (Ansage des Nutzers, 08/2026: „Mir fehlt bei allen
/// Elementen, die automatisch ablaufen, die Funktion, sie wieder auf einen
/// unbenutzten Zustand zurückzustellen.").
///
/// **Was zurückgesetzt wird, ist der Ablauf — nie die Einrichtung.** Die
/// Namensliste bleibt, der Grundriss bleibt, die Dauer des Timers bleibt,
/// Farben und Beschriftungen bleiben. Und **Archive bleiben**: die
/// gesicherten Ziehungen und die gesicherten Sitzordnungen sind ein
/// Nachweis, kein Zustand. Wer sie loswerden will, löscht sie dort, wo sie
/// stehen.
extension WidgetContent {

    /// Steht an diesem Element etwas, das ein Zurücksetzen entfernen würde?
    ///
    /// Danach richtet sich, ob der Menüpunkt überhaupt etwas bewirkt — ein
    /// Knopf, der nichts tut, ist schlimmer als keiner.
    var benutzt: Bool {
        switch self {
        case .namePicker(let c):
            return c.currentID != nil || !c.drawnIDs.isEmpty || !c.ergebnis.isEmpty
                || !c.erledigt.isEmpty || !c.revealParts.isEmpty || c.festgehalten
                || !c.zaehler.isEmpty || !c.paare.isEmpty
        case .timer(let c):
            return c.endsAtMs != nil || c.startedAtMs != nil || c.pausedValue != nil
        case .trafficLight(let c):
            return c.state != .green
        case .checklist(let c):
            return c.items.contains { $0.done }
        case .kamera(let c):
            return c.eingefroren != nil
        case .geburtstag(let c):
            return c.ritual > 0 || !c.gratulanten.isEmpty || !c.rollen.isEmpty
                || !c.fragen.isEmpty
        case .sitzplan(let c):
            return !c.belegung.isEmpty || c.aufgedeckt > 0 || c.gesperrt
                || !c.bericht.isEmpty
        case .clock, .noise, .text, .image, .sounds, .symbols, .video:
            // Diese laufen nicht ab — eine Uhr ist nie „benutzt".
            return false
        }
    }

    /// Dasselbe Element im unbenutzten Zustand.
    var unbenutzt: WidgetContent {
        switch self {
        case .namePicker(var c):
            c.currentID = nil
            c.drawnIDs = []
            c.ergebnis = []
            c.erledigt = []
            c.revealParts = []
            c.festgehalten = false
            // Das Gedächtnis gehört zum Gebrauch: Wer wie oft gezogen wurde
            // und wer mit wem in einer Gruppe war, ist die Spur der letzten
            // Wochen. `ziehungen` bleibt — das ist das Archiv.
            c.zaehler = [:]
            c.paare = [:]
            c.ziehungID = ""
            return .namePicker(c)

        case .timer(var c):
            c.endsAtMs = nil
            c.startedAtMs = nil
            c.pausedValue = nil
            return .timer(c)

        case .trafficLight(var c):
            c.state = .green
            return .trafficLight(c)

        case .checklist(var c):
            for stelle in c.items.indices { c.items[stelle].done = false }
            return .checklist(c)

        case .kamera(var c):
            c.eingefroren = nil
            return .kamera(c)

        case .geburtstag(var c):
            c.ritual = 0
            c.gratulanten = []
            c.rollen = []
            c.fragen = []
            return .geburtstag(c)

        case .sitzplan(var c):
            c.belegung = [:]
            c.namen = [:]
            c.reihenfolge = []
            c.aufgedeckt = 0
            c.bericht = []
            c.gesperrt = false
            c.laufenderEintrag = ""
            // `plaetze` und `archiv` bleiben: der Grundriss ist Einrichtung,
            // das Archiv ist ein Nachweis.
            return .sitzplan(c)

        case .clock, .noise, .text, .image, .sounds, .symbols, .video:
            return self
        }
    }
}

File: klassenraum/js/zuruecksetzen.js
// Auf unbenutzt zurücksetzen — übernommen aus der Tafelbild-App (1.3.21/1.3.24).
//
// Während der Stunde ist es genau richtig, dass der gezogene Name, die
// Sitzordnung und die gelaufene Feier stehen bleiben — eine Tafel, die sich
// selbst aufräumt, während die Klasse noch hinsieht, wäre unbrauchbar. Am
// nächsten Morgen ist es falsch. Zurück geht der GEBRAUCH, nie die
// Einrichtung (Namensliste, Grundriss, Dauer, Farben) und nie die Archive:
// Die sind ein Nachweis, kein Zustand.
//
// Zwei Tiefen (nur beim Zufälligen Namen verschieden):
// - 'ergebnis' (Vorgabe): die Tafel zeigt wieder die Ansicht vor der
//   Auslosung, weiß aber weiter, wer schon dran war. Wer am Montag
//   zurücksetzt und das Gedächtnis verliert, zieht am Dienstag womöglich
//   dasselbe Kind zum dritten Mal.
// - 'alles': dazu das Gedächtnis (Gezogene, Wer-mit-wem, Wer-war-dran).

/** Hat dieses Element einen Ablauf, der sich zurücksetzen lässt — und ist
 *  gerade etwas davon in Gebrauch? Ein Knopf, der nichts tut, ist schlimmer
 *  als keiner. */
export function istBenutzt(widget, tiefe = 'ergebnis') {
  const state = widget.state || {};
  switch (widget.type) {
    case 'randomizer': {
      const ergebnis = Boolean(state.current) || Boolean(state.groups)
        || (state.revealParts || []).length > 0 || state.locked === true;
      if (tiefe === 'ergebnis') return ergebnis;
      return ergebnis || (state.drawn || []).length > 0
        || Object.keys(state.paare || {}).length > 0
        || Object.keys(state.dran || {}).length > 0;
    }
    case 'timer':
      return Boolean(state.running) || Boolean(state.startedAt)
        || (state.mode === 'timer'
          ? state.remaining !== state.seconds
          : (state.elapsed || 0) > 0);
    case 'traffic':
      return state.active !== 'green';
    case 'checklist':
      return (state.items || []).some((item) => item.done);
    case 'camera':
      return Boolean(state.frozen);
    case 'birthday':
      return !state.hinweis && (state.ritual || 0) > 0;
    case 'seating':
      return Object.keys(state.belegung || {}).length > 0 || state.gesperrt === true;
    default:
      // Uhr, Lautstärke, Text, Bild, Klänge, Arbeitssymbol und Video laufen
      // nicht ab — eine Uhr ist nie „benutzt".
      return false;
  }
}

/** Gibt es für diese Art überhaupt eine tiefere Stufe? Nur beim Zufälligen
 *  Namen — überall sonst wären beide Tiefen dasselbe, und zwei
 *  gleichbedeutende Knöpfe sind schlimmer als einer. */
export function hatGedaechtnis(widget) {
  return widget.type === 'randomizer';
}

/** Den Gebrauch eines Elements zurücknehmen. Liefert true, wenn sich etwas
 *  geändert hat. */
export function setzeZurueck(widget, tiefe = 'ergebnis') {
  if (!istBenutzt(widget, tiefe)) return false;
  const state = widget.state || {};
  switch (widget.type) {
    case 'randomizer':
      state.current = null;
      state.groups = null;
      state.revealParts = [];
      state.locked = false;
      if (tiefe === 'alles') {
        state.drawn = [];
        state.paare = {};
        state.dran = {};
      }
      return true;
    case 'timer':
      state.running = false;
      state.endsAt = null;
      state.startedAt = null;
      state.remaining = state.seconds;
      state.elapsed = 0;
      return true;
    case 'traffic':
      state.active = 'green';
      return true;
    case 'checklist':
      (state.items || []).forEach((item) => { item.done = false; });
      return true;
    case 'camera':
      state.frozen = '';
      return true;
    case 'birthday':
      state.ritual = 0;
      state.gratulanten = [];
      state.rollen = [];
      state.fragen = [];
      return true;
    case 'seating':
      state.belegung = {};
      state.reihenfolge = [];
      state.aufgedeckt = 0;
      state.bericht = [];
      state.gesperrt = false;
      state.laufenderEintrag = '';
      return true;
    default:
      return false;
  }
}

/** Alle Elemente einer Seite zurücksetzen. Liefert, wie viele sich geändert
 *  haben. */
export function seiteZuruecksetzen(page, tiefe = 'ergebnis') {
  let getan = 0;
  for (const widget of page.widgets || []) {
    if (setzeZurueck(widget, tiefe)) getan += 1;
  }
  return getan;
}

/** Die ganze Tafel zurücksetzen. */
export function tafelZuruecksetzen(board, tiefe = 'ergebnis') {
  let getan = 0;
  for (const page of board.pages || []) {
    getan += seiteZuruecksetzen(page, tiefe);
  }
  return getan;
}

/** Ist auf der Seite / der Tafel überhaupt etwas zu vergessen? */
export function seiteBenutzt(page, tiefe = 'ergebnis') {
  return (page.widgets || []).some((widget) => istBenutzt(widget, tiefe));
}

export function tafelBenutzt(board, tiefe = 'ergebnis') {
  return (board.pages || []).some((page) => seiteBenutzt(page, tiefe));
}
